### PR TITLE
167 plain bundle validation

### PR DESCRIPTION
**Pull request by asmacdo**
Add plain bundle v0 struct, use initialization for validation
Closes https://github.com/operator-framework/rukpak/issues/167

### CODE REVIEW

File: cmd/core/main.go [hunk 1]
```diff
@@ -205,12 +206,19 @@ func main() {
 		os.Exit(1)
 	}
 
+	plainValidator, err := validators.NewPlainValidator()
+	if err != nil {
+		setupLog.Error(err, "unable to setup bundle validator")
+		os.Exit(1)
+	}
```
Comment (joelanford) [inline cmd/core/main.go:213]: Do we plan to use a different validator with the plain provisioner? If not, I'd suggest hiding this as an implementation detail of the plain bundle reconciler rather than making it a configurable thing on the struct.

File: cmd/core/main.go [hunk 2]
```diff
@@ -205,12 +206,19 @@ func main() {
 		os.Exit(1)
 	}
 
+	plainValidator, err := validators.NewPlainValidator()
```
Comment (joelanford) [inline cmd/core/main.go:209]: IMO, anything specific to the plain provisioner should go in the `internal/proivisioner/plain` tree somewhere.

File: internal/validators/base.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+	apimachyaml "k8s.io/apimachinery/pkg/util/yaml"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Validator interface {
+	Validate(fs.FS) error
+}
+
+// TODO(asmacdo) I moved this *and* made it public, is this a problem?
+func GetObjects(bundleFS fs.FS) ([]client.Object, error) {
```
Comment (joelanford) [inline internal/validators/base.go:21]: Do we use this anywhere outside the plain provisioner (I think not since it has knowledge and makes assumptions about the FS that's passed in)?

If it is specific to the plain provisioner, I think it should be internal and within the plain package tree.

File: internal/validators/base.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	"path/filepath"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	apimachyaml "k8s.io/apimachinery/pkg/util/yaml"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Validator interface {
+	Validate(fs.FS) error
+}
```
Comment (joelanford) [inline internal/validators/base.go:18]: This might be a premature interface definition. I'd expect to see this crop up only when there's a use case around passing different validator implementations around to the same provisioner.

File: internal/validators/plain.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+	validationErrors = append(validationErrors, errors.New("hello"))
+	validationErrors = append(validationErrors, errors.New("world"))
+	validationErrors = append(validationErrors, errors.New("annnndd"))
+	validationErrors = append(validationErrors, errors.New("four Errors!"))
+	validationErrors = append(validationErrors, errors.New("moved this shit around!"))
+
+	err = errors.New("")
+	for _, each := range validationErrors {
+		err = fmt.Errorf("%v\n%v", err, each)
+	}
```
Comment (joelanford) [inline internal/validators/plain.go:39]: This seems like it will always result in a non-nil error, which seems undesirable. Perhaps consider using https://pkg.go.dev/k8s.io/apimachinery/pkg/util/errors#Aggregate

Comment (asmacdo) [reply internal/validators/plain.go:39]: ah thanks, I was looking for errors.Wrap and didn't see it

File: cmd/core/main.go (hunk 1 above)
Comment (asmacdo) [reply cmd/core/main.go:213]: I wasn't sure, but I imagined but I figured the registry provisioner (or any other "conversion" provisioners) would use this as well.

I may have over engineered, but given the recent history of operator-lib validations I am imagining that the reconcilers may need to use different versions of the same validators, or that reconcilers might share validators.

**Review by timflannagan — COMMENTED** [review]
Just did a quick run through of these changes, and I mainly had the same feedback as Joe. I think my preference would be to avoid making the first pass implementation here generalistic, and instead just focusing on the plain provisioner use case without exposing too much of this through configuration.

In terms of exposing a knob on the plain provisioner for configuring validators: would we need a way for admins to opt-out to this validation? Outside of insufficient testing for our validation logic, is there any way that an admin would want/need to be able to opt-out of these validations in the future?

File: internal/validators/base.go (hunk 3 above)
Comment (asmacdo) [reply internal/validators/base.go:21]: bundle deployment uses it as well https://github.com/operator-framework/rukpak/blob/0f45b79342c53aab4e1ee4dee1958f7c95e3e422/internal/provisioner/plain/controllers/bundledeployment_controller.go#L436

**Comment by tylerslaton** [comment]
@asmacdo Another thing, could we include a `Closes #167` in the description so that automation closes the issue out when this merges?

**Comment by asmacdo** [comment]
@tylerslaton it's in the commit message, does "Fixed" work also?

**Comment by tylerslaton** [comment]
> @tylerslaton it's in the commit message, does "Fixes" work also?

Ah, I see what you're saying. I think that the [automation](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword#linking-a-pull-request-to-an-issue-using-a-keyword) should detect the linking if you remove the `operator-framework` prefix or change it to be `operator-framework/rukpak#167`. You'll know if it works when the issue is linked in the `Development` section of the right hand bar.

File: cmd/core/main.go (hunk 1 above)
Comment (joelanford) [reply cmd/core/main.go]: > I imagined the registry provisioner (or any other "conversion" provisioners) would use this as well.

I could see that as a sort of post-conversion step perhaps, but if our registry+v1 to plain+v0 conversion resulted in a failure in the plain bundle validation, that'd be a bug on our side, not really something that would be useful/actionable for a user.

Perhaps it's worth structuring it such that the registry+v1 to plain+v0 conversion tests can call it as a sanity check?

**Comment by joelanford** [comment]
> would we need a way for admins to opt-out to this validation? 

I'd imagine that the level of validation we should do here is "will the bundle deployment provisioner fail to parse this bundle? If yes, validation error. If no, no error". In that case, there's no reason to give cluster admins a knob for it.

If we ever get into more linting kinds of validations, we would want cluster admin knobs. But IMO linting shouldn't be done at runtime.

File: internal/provisioner/plain/controllers/bundle_controller.go [hunk 6]
```diff
@@ -229,6 +223,32 @@ func updateStatusUnpackFailing(u *updater.Updater, err error) error {
 	return err
 }
 
+func (r *BundleReconciler) validate(bundle fs.FS) error {
+	var validationErrors []error
+	objects, err := getObjects(bundle)
+	if err != nil {
+		validationErrors = append(validationErrors, fmt.Errorf("get objects from bundle manifests: %v", err))
+	}
+	if len(objects) == 0 {
```
Comment (joelanford) [inline internal/provisioner/plain/controllers/bundle_controller.go:232]: `else if` here?

I assume that if we encounter an error getting the objects, the length of `objects` will always be 0.

File: internal/provisioner/plain/validator/validator.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+	if err != nil {
+		validationErrors = append(validationErrors, fmt.Errorf("get objects from bundle manifests: %v", err))
+	} else if len(objects) == 0 {
+		validationErrors = append(validationErrors, errors.New("invalid bundle: found zero objects: "+
+			"plain+v0 bundles are required to contain at least one object"))
+	}
+	return apimacherrors.NewAggregate(validationErrors)
+}
+
+func GetObjects(bundleFS fs.FS) ([]pkgclient.Object, error) {
```
Comment (joelanford) [inline internal/provisioner/plain/validator/validator.go:29]: Seems a bit odd to me that `GetObjects` is in the `validate` package. Perhaps we should have a `internal/bundle/plain/` package that contains a `LoadBundleV0(fs.FS) (*BundleV0, error)` function.

And then all of the logic for enforcing the plain+v0 bundle spec would be contained there (and could still be abstracted out into a separate set of validations functions)?

**Comment by asmacdo** [comment]
@joelanford I attempted to use the cli-runtime library to serialize and validate the json/yaml. I was able to get this to work for validation purposes.

However, this turned out to be a dead end (AFAICT) because the [`Info.Objects`](https://github.com/kubernetes/cli-runtime/blob/v0.25.0/pkg/resource/visitor.go#L82) are `k8s.io/apimachinery/pkg/runtime.Objects` rather than `sigs.k8s.io/controller-runtime/pkg/client.Objects`. The bundle deployment controller uses `client.Object.GetLabels()` which is not implemented in the `runtime.Object`

Therefore, I've opted to keep the `getObjects` code (renamed and moved to `LoadV0`) rather than refactor it to use the cli-runtime library.

**Comment by joelanford** [comment]
@asmacdo, we can type assert those `runtime.Object`s to `client.Object`s.

File: internal/bundle/plain.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+			return PlainV0{}, fmt.Errorf("subdirectories are not allowed within the %q directory "+
+				"of the bundle image filesystem: found %q", manifestsDir, filepath.Join(manifestsDir, manifest.Name()))
+		}
+		manifestReader, err := bundle.Open(filepath.Join("manifests", manifest.Name()))
+		if err != nil {
+			return PlainV0{}, err
+		}
+		defer manifestReader.Close()
+		// Stream().Do() visits and serializes each object
+		result := resource.NewLocalBuilder().Flatten().Unstructured().Stream(manifestReader, "Load PlainV0 Bundle").Do()
```
Comment (joelanford) [inline internal/bundle/plain.go:38]: Rather than using `"Load PlainV0 Bundle"`, I think the intention of that string is to be something descriptive about the `manifestReader`. I'd suggest extracting `filepath.Join("manifests", manifest.Name())` as a variable and using that when we `Open` and here.

File: internal/bundle/plain.go [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+	for _, manifest := range dirEntries {
+		if manifest.IsDir() {
+			return PlainV0{}, fmt.Errorf("subdirectories are not allowed within the %q directory "+
+				"of the bundle image filesystem: found %q", manifestsDir, filepath.Join(manifestsDir, manifest.Name()))
+		}
+		manifestReader, err := bundle.Open(filepath.Join("manifests", manifest.Name()))
+		if err != nil {
+			return PlainV0{}, err
+		}
+		defer manifestReader.Close()
```
Comment (joelanford) [inline internal/bundle/plain.go:36]: Nit: `defer`-ing in a loop results in all of the deferred functions running at the end of the function rather than immediately after we've consumed that specific file. Can we extract this loop code into a separate function, basically:
```go
for _, manifest := range dirEntries {
	manifestObjects, err := getObjects(bundle, manifest)
	if err != nil {
		return nil, err
	}
	objects = append(objects, manifestObjects...)
}
```

That way, when you `defer` in `getObjects`, you end up actually closing the file before moving on to the next file.

Comment (joelanford) [reply internal/bundle/plain.go:36]: We could also potentially do 
```go
fs.WalkDir(bundle, manifestsDir, func(path string, d fs.DirEntry, err error) error {
	// handle individual dir entry in this closure.
})
```

File: internal/bundle/plain_test.go [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+				Data: []byte("kind: ofnecessary"),
+			},
+			"manifests/invalid/valid.yaml": {
+				Data: []byte("kind: illegalsubdir"),
+			},
+		}
+		_, err := LoadPlainV0(mockfs)
+		Expect(err.Error()).To(ContainSubstring("subdirectories are not allowed"))
+	})
+	It("should pass with multi-object yaml files", func() {
```
Comment (joelanford) [inline internal/bundle/plain_test.go:87]: Can you also add tests for:
1. All manifest files contain JSON
2. Manifest files are a mix of JSON and YAML (but each file is one or the other)
3. The ability to parse and flatten `v1.List` objects that contain a list of `items` that are objects.

File: internal/provisioner/plain/controllers/bundledeployment_controller.go [hunk 11]
```diff
@@ -370,14 +371,13 @@ func (r *BundleDeploymentReconciler) loadBundle(ctx context.Context, bundle *ruk
...
+	plainV0, err := bundleLib.LoadPlainV0(bundleFS)
 	if err != nil {
 		return nil, fmt.Errorf("read bundle objects from bundle: %v", err)
 	}
 
-	objs := make([]client.Object, 0, len(objects))
-	for _, obj := range objects {
-		obj := obj
+	objs := make([]client.Object, 0, len(plainV0.Objects))
+	for _, obj := range plainV0.Objects {
```
Comment (joelanford) [inline internal/provisioner/plain/controllers/bundledeployment_controller.go:380]: Any reason we can't/shouldn't modify the `plainV0.Objects` slice directly? It would use less memory if we modified the objects in place.

**Review by joelanford — COMMENTED** [review]
Looks good overall. Just left a handful of suggestions for minor improvements.

File: internal/bundle/plain_test.go [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+
+import (
+	"io/fs"
+	"testing/fstest"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Plain bundle v0 loader", func() {
```
Comment (tylerslaton) [inline internal/bundle/plain_test.go:11]: @joelanford @timflannagan Do we have a preference for using Ginkgo vs Go's standard testing framework when unit testing? We use Go's standard framework for some instance and Ginkgo for others .

Comment (joelanford) [reply internal/bundle/plain_test.go:11]: I personally prefer ginkgo, but I realize its a contentious debate. I'm more interested in consistency and ease of adding new test cases to existing test suites than I am about any particular framework.

File: internal/provisioner/plain/controllers/bundledeployment_controller.go (hunk 11 above)
Comment (asmacdo) [reply internal/provisioner/plain/controllers/bundledeployment_controller.go:380]: at some point i was doing the type assert here, this was just an artifact of that. good catch!

File: internal/bundle/plain.go (hunk 8 above)
Comment (joelanford) [reply internal/bundle/plain.go]: Looks like this one got lost in the latest push?

File: internal/bundle/plain_test.go [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+		Expect(err).To(BeNil())
+	})
+	It("should pass if the JSON is valid and contains a `kind` element", func() {
+		mockfs := fstest.MapFS{
+			"manifests/valid-json.yaml": {
+				Data: []byte(`{"kind": "ofnecessary"}`),
+			},
+			"manifests/valid.yaml": {
+				Data: []byte("kind: ofnecessary"),
+			},
```
Comment (joelanford) [inline internal/bundle/plain_test.go:72]: Seems like this test is identical to the one below. Looks like this is supposed to test a JSON file by itself.
```suggestion
			"manifests/valid.json": {
				Data: []byte(`{"kind": "ofnecessary"}`),
			},
```

File: internal/bundle/plain_test.go [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+	It("should flatten and pass with metav1.List yaml files", func() {
+		mockfs := fstest.MapFS{
+			"manifests/valid.yaml": {
+				Data: []byte("kind: List\napiVersion: v1\nitems:\n  - kind: me\n    apiVersion: v1\n  - kind: you\n    apiVersion: v1"),
+			},
+		}
+		plainv0, err := LoadPlainV0(mockfs)
+		Expect(len(plainv0.Objects)).To(Equal(2))
+		Expect(err).To(BeNil())
+	})
```
Comment (joelanford) [inline internal/bundle/plain_test.go:123]: Another potential test: 2 files, one is valid YAML, the other is empty.

From the code that exists right now, that looks valid. But I wonder if we should call that invalid (i.e. every file in the manifests dir MUST have at least one object)

File: internal/provisioner/plain/controllers/bundledeployment_controller.go [hunk 15]
```diff
@@ -46,6 +46,7 @@ import (
 	"sigs.k8s.io/yaml"
 
 	rukpakv1alpha1 "github.com/operator-framework/rukpak/api/v1alpha1"
+	bundleLib "github.com/operator-framework/rukpak/internal/bundle"
```
Comment (joelanford) [inline internal/provisioner/plain/controllers/bundledeployment_controller.go:49]: nit: drop the import alias and just use the `bundle` package name in the code.

if i'm missing another import already called `bundle` that requires this alias, i'd suggest `bundlelib` all lowercase. `bundleLib` makes mind mind drift to "that's a variable" rather than "that's a package".

File: internal/bundle/plain.go [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+		objects = append(objects, *theseObjects...)
+	}
+	if len(objects) == 0 {
+		return nil, errors.New("invalid bundle: found zero objects: " +
+			"plain+v0 bundles are required to contain at least one object")
+	}
+	return &PlainV0{Objects: objects}, nil
+}
+
+func getObjects(bundle fs.FS, manifest fs.DirEntry) (*[]client.Object, error) {
```
Comment (joelanford) [inline internal/bundle/plain.go:45]: No need to return a pointer to a slice. A slice is already a pointer (so you can continue returning `nil, err`).
```suggestion
func getObjects(bundle fs.FS, manifest fs.DirEntry) ([]client.Object, error) {
```

File: internal/bundle/plain.go [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+	manifestReader, err := bundle.Open(filepath.Join("manifests", manifest.Name()))
```
Comment (timflannagan) [inline internal/bundle/plain.go:46]: nit: avoid hardcoding the "manifests" directory and instead use a constant?

File: internal/bundle/plain.go [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+	}
+	infos, err := result.Infos()
+	if err != nil {
+		return nil, err
+	}
+	objects := make([]client.Object, 0, len(infos))
+	for _, info := range infos {
+		clientObject := info.Object.(client.Object)
+		objects = append(objects, clientObject)
+	}
```
Comment (timflannagan) [inline internal/bundle/plain.go:64]: You might be able to modularize this better, and move this logic into it's own function that returns an array of client.Object(s).

File: internal/bundle/plain_test.go [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Plain bundle v0 loader", func() {
+	It("should fail if there is no manifests directory", func() {
+		mockfs := fstest.MapFS{}
+		_, err := LoadPlainV0(mockfs)
+		Expect(err.Error()).To(ContainSubstring("file does not exist"))
```
Comment (timflannagan) [inline internal/bundle/plain_test.go:15]: It looks like we're fairly consistent with using `Expect(err).To(MatchError(ContainSubstring(...))` throughout ginkgo-related tests that perform this kind of testing check.

File: internal/bundle/plain.go [hunk 20]
```diff
@@ -0,0 +1,66 @@
+package bundle
```
Comment (timflannagan) [inline internal/bundle/plain.go:1]: This might be a good candidate for it's own PR given we're moving around the getObjects function here.

**Review by timflannagan — COMMENTED** [review]
Quick review. Looks like we're heading in the right direction.

File: internal/bundle/plain.go (hunk 8 above)
Comment (asmacdo) [reply internal/bundle/plain.go]: nice catch thanks

File: internal/provisioner/plain/controllers/bundledeployment_controller.go (hunk 15 above)
Comment (asmacdo) [reply internal/provisioner/plain/controllers/bundledeployment_controller.go:49]: actually BundleReploymentReconciler.loadBundle has `bundle` as an arg. IMO that param name makes sense, so I'm keeping `bundleLib`

File: internal/bundle/plain.go [hunk 21]
```diff
@@ -14,8 +14,9 @@ type PlainV0 struct {
 	Objects []client.Object
 }
 
+const manifestsDir = "manifests"
```
Comment (joelanford) [inline internal/bundle/plain.go:17]: nit: this is a variable that applies to the whole bundle package, so if its going to be a global constant, i'd name it something like `plainv0ManifestsDir`

File: internal/bundle/plain_test.go [hunk 22]
```diff
@@ -0,0 +1,121 @@
...
+
+import (
+	"io/fs"
+	"testing/fstest"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Plain bundle v0 loader", func() {
```
Comment (joelanford) [inline internal/bundle/plain_test.go:11]: Nit: pretty subjective, but when a Describe is clearly scoped to a particular function or type, I typically use the function name as the describe string.
```go
var _ = Describe("LoadPlainV0", func() { ... })
```

**Comment by tylerslaton** [comment]
@asmacdo Mind rebasing this? I think this is likely ready to go in unless @timflannagan has any outstanding comments.

**Comment by asmacdo** [comment]
Joes refactor seems to have done this generally, so no need to do it for plain bundles like this.
